### PR TITLE
Implement SaveChunkEvent.

### PR DESCRIPTION
**Pull request by Zidane**
For discussion, @dualspiral 

Also note I fixed the fact that we were calling LoadChunk on the client thread (or attempting to) as well as we were not respecting ShouldFire for unload.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

### CODE REVIEW

**Comment by dualspiral** [comment]
I mean, it's bog standard throwing events and fixing bugs! I see nothing contentious here. If someone does, I'd be mightly surprised. When the API is good to go, this will be good to go.

File: src/main/java/org/spongepowered/common/mixin/core/world/chunk/storage/AnvilChunkLoaderMixin.java [hunk 1]
```diff
@@ -300,4 +305,11 @@ public boolean writeNextIO() {
         return this.chunkSaveLocation.toPath();
     }
 
-}
+    @Inject(method = "writeChunkData", at = @At("RETURN"))
+    private void impl$callSaveChunkEventPost(ChunkPos pos, NBTTagCompound compound, CallbackInfo ci) {
+        if (ShouldFire.SAVE_CHUNK_EVENT || ShouldFire.SAVE_CHUNK_EVENT_POST) {
```
Comment (ImMorpheus) [inline src/main/java/org/spongepowered/common/mixin/core/world/chunk/storage/AnvilChunkLoaderMixin.java]:
```suggestion
        if (ShouldFire.SAVE_CHUNK_EVENT_POST) {
```

File: src/main/java/org/spongepowered/common/mixin/core/world/gen/ChunkProviderServerMixin.java [hunk 2]
```diff
@@ -325,4 +326,15 @@ public boolean tick()
         this.loadedChunks.remove(ChunkPos.asLong(chunk.x, chunk.z));
         ((ChunkBridge) chunk).bridge$setScheduledForUnload(-1);
     }
+
+    @Inject(method = "saveChunkData", at = @At("HEAD"), cancellable = true)
+    private void impl$callSaveChunkEventPre(Chunk chunkIn, CallbackInfo ci) {
+        if (ShouldFire.SAVE_CHUNK_EVENT || ShouldFire.SAVE_CHUNK_EVENT_PRE) {
```
Comment (ImMorpheus) [inline src/main/java/org/spongepowered/common/mixin/core/world/gen/ChunkProviderServerMixin.java]:
```suggestion
        if (ShouldFire.SAVE_CHUNK_EVENT_PRE) {
```

**Review by ImMorpheus — CHANGES_REQUESTED** [review]
Checking the flag for the generic type will also make the event fire when there are listeners for other sub events but not the event itself.

**Review by ImMorpheus — APPROVED** [review]

**Review by gabizou — APPROVED** [review]
LGTM. Some reservations about changing `world.isRemote` to a `((WorldBridge) world).bridge$isFake()` but that isn't likely going to be an issue.

**Pull request merged**